[addons] use on CAddonDll instance class pointer as identifier

AlwinEsch opened the pull request:
## Description

Before was a string used (where was in most cases only a pointer as string).
Seen during PVR rework, where it becomes needed to use a own identifier, to pass
same name several times to identify e.g. "channel" or "recording".

With same string then comes in conflict with the std::map where store used
Instances on CAddonDll, by use of the related class pointer this is more safe.

Only was related to game addon system, where currently not use
`IAddonInstanceHandler` (games to complex for me to find direct a way with them),
a add of `PSEUDO_INSTANCE_CLASS` (const void*) needed.

On addon interface change this nothing, only store the map in Kodi it by
different value.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

peak3d commented on xbmc/addons/binary-addons/AddonDll.h:
```diff
@@ -118,7 +124,7 @@ namespace ADDON
     bool m_initialized;
     bool LoadDll();
-    std::map<std::string, std::pair<ADDON_TYPE, KODI_HANDLE>> m_usedInstances;
+    std::unordered_map<PSEUDO_INSTANCE_CLASS, std::pair<ADDON_TYPE, KODI_HANDLE>> m_usedInstances;
```
This affects performance when calling find(), for what is the change necessary?

AlwinEsch replied:
I know, but it have the find only during destruction of the instance (where it have more time), during creation of instance and insert for there becomes the sort prevented and added on back.
It does not change much, thought only why not and faster is faster.

But is not a place where we can speak about much entries, currently I think max. 3 entries, maybe in future to 10.

peak3d replied:
ok

peak3d replied:
But: inserting is also slower, because std::map has to find for duplicates first before insertion occurs. I know it does not change much, but I don't see any reason for changing it to unordered.

AlwinEsch replied:
Commit becomes removed :simple_smile:

peak3d commented on xbmc/addons/binary-addons/AddonDll.cpp:
```diff
@@ -317,7 +317,11 @@ void CAddonDll::Destroy()
 
-ADDON_STATUS CAddonDll::CreateInstance(ADDON_TYPE instanceType, const std::string& instanceID, KODI_HANDLE instance, KODI_HANDLE parentInstance)
+ADDON_STATUS CAddonDll::CreateInstance(ADDON_TYPE instanceType,
+                                       PSEUDO_INSTANCE_CLASS instanceClass,
```
Why is this new param prefixed with "PSEUDO" ? I don't really get the meaning for it

AlwinEsch replied:
Is thought to replace after the Game addon use `IAddonInstanceHandler`, as long that have CAddonDll for there, was it only a Pseudo for me :smile:.

peak3d reviewed (commented):
I read this PR like a cosmetic change, maybe a little bit improvement because we search for void* insttead string in the map. Can you please explain how this PR improves anything? Or if it really fixes anything? Instead of std::string an size_t would be also fine as and identifier (??)

AlwinEsch commented on xbmc/addons/binary-addons/AddonDll.h:
```diff
@@ -18,6 +18,7 @@ namespace ADDON
 {
 
   typedef void* (*ADDON_GET_INTERFACE_FN)(const std::string &name, const std::string &version);
+  typedef const void* PSEUDO_INSTANCE_CLASS;
```
Becomes changed, thanks!

AlwinEsch commented:
Is not direct a fix, but a big prevent of possible problems, I seen it by change of PVR, If CAddonDll/IAddonInstanceHandler becomes from outside a identify string can it be, that this is already used, if this then stored inside the `map` becomes the one before overwritten. 

Thats a very silent bug and can generate some mem leaks on addon with missing destruction calls to them.

AlwinEsch committed:
[addons] use on CAddonDll instance class pointer as identifier

Before was a string used (where was in most cases only a pointer as string).
Seen during PVR rework, where it becomes needed to use a own identifier, to pass
same name several times to identify e.g. "channel" or "recording".

With same string then comes in conflict with the std::map where store used
Instances on CAddonDll, by use of the related class pointer this is more safe.

Only was related to game addon system, where currently not use
IAddonInstanceHandler (games to complex for me to find direct a way with them),
a add of ADDON_INSTANCE_HANDLER (const void*) needed.

On addon interface change this nothing, only store the map in Kodi it by
different value.

AlwinEsch commented:
Have updated, the one with "unordered_map" is removed and added a new second commit to change "xbmc/addons/binary-addons/AddonDll.h" to match the code guidelines (to have for from now correct).

AlwinEsch committed:
[addons] Code guideline update for xbmc/addons/binary-addons/AddonDll.h

garbear reviewed (approved):
Nice improvement
